File: 2601-prime-subtraction-operation/2601-prime-subtraction-operation.js
/**
 * @param {number[]} nums
 * @return {boolean}
 */
var primeSubOperation = function(nums) {
    const pick = (num, r) => {
        let p = Math.max(2, num - r + 1), d
        if (p >= num)
            return 0
        if (p == 2)
            return 2
        if (p % 2 == 0)
            p += 1
        while (p < num) {
            d = 3
            while (p % d) {
                if (p < d * d)
                    return p
                d += 2
            }
            p += 2
        }
        return 0
    }
    let d = nums.at(-1)
    for (let i = nums.length-2; i >= 0; i--) {
        if (nums[i] >= d) {
            p = pick(nums[i], d)
            if (p == 0)
                return false
            d = nums[i] - p
        } else 
            d = nums[i]
    }
    return true
};

File: 2601-prime-subtraction-operation/2601-prime-subtraction-operation.py
class Solution:
    def primeSubOperation(self, nums: List[int]) -> bool:
        prime=[True]*1001
        prime[0]=prime[1]=False
        for x in range(2,1001):
            if prime[x]:
                for i in range(x*x,1001,x):
                    prime[i]=False


        prev=0
        for x in nums:
            if prev>=x:
                return False

            for p in range(x-1,-1,-1):
                if prime[p] and x-p>prev:
                    break

            prev=x-p

        return True                            
        
        